cwrc_audit_report.rb

```diff
@@ -0,0 +1,144 @@
+#     Status
+#
+#     where:
+#       status =
+#          if 'x' then needs preservation
+#           else if 'd' then not present within CWRC
+#           else if 'x' then Swift object is of zero size
+#           else '' then ok
+#
+# Usage: <progname> [options]...
+#   options
+#     -h --help
+#     -s --summary summary output where status in not 'ok'
+#
+
+require 'logger'
+require 'optparse'
+require 'time'
+require 'swift_ingest'
+
+require_relative 'cwrc_common'
+
+module CWRCPerserver
+  # status IDs
+  STATUS_OK = ''.freeze
+  STATUS_E_SIZE = 's'.freeze # error: size zero or too small
+  STATUS_I_FLAG = 'x'.freeze # flagged for preservation
+  STATUS_I_DEL = 'd'.freeze # missing from the CWRC side while Swift contains a copy
+
+  opt_summary_output = false
+  ARGV.options do |opts|
+    opts.on('-s', '--summary', "Summary output where status is not 'ok'") do
+      opt_summary_output = true
+    end
+    opts.on_tail('-h', '--help', 'Displays help') do
+      puts opts
+      exit
+    end
+    opts.parse!
+  end
+
+  # initialize environment
+  set_env
+
+  # authenticate to CWRC repository
+  cookie = retrieve_cookie
+
+  # query the CWRC repository
+  # response: {"pid"=>"cwrc:c1583789-0dad-41d3-8a42-94d7a8e6d451", "timestamp"=>"2018-05-02T17:07:29.028Z"}
+  cwrc_objs = get_cwrc_objs(cookie, '')
+
+  # connect to swift storage
+  swift_con = connect_to_swift
+  raise CWRCArchivingError if swift_con.nil?
+
+  # query Swift storage for a list of objects
+  # https://github.com/ruby-openstack/ruby-openstack/wiki/Object-Storage
+  # https://github.com/ruby-openstack/ruby-openstack/wiki/Object-Storage
+  # response: "cwrc_0c168793-b1ff-453f-a1f6-e1d75f7350be"=>{
+  #    :bytes=>"5939",
+  #    :content_type=>"application/x-tar",
+  #    :last_modified=>"2018-02-05T06:45:23.422720",
+  #    :hash=>"dd2b11f239f7f25fb504519b612cf896"
+  #  },
+  swift_container = swift_con.swift_connection.container(swift_con.project)
+
+  # Iterate via markers
+  # https://github.com/ruby-openstack/ruby-openstack/blob/d9c8aa19488062e483771a9168d24f2626fe688b/lib/openstack/swift/container.rb#L100
+  swift_objs = swift_container.objects_detail
+  while swift_objs.count < swift_container.container_metadata[:count].to_i
+    swift_objs = swift_objs.merge(swift_container.objects_detail(marker: swift_objs.keys.last))
+  end
+
+  # TODO: use CSV gem
+  # CSV header
+  puts "cwrc_pid (#{cwrc_objs.count}),"\
+    "cwrc_mtime (#{Time.now}),"\
+    "swift_id (#{swift_container.container_metadata[:count]}),"\
+    'swift_timestamp,swift_bytes,status'
+
+  # TODO: find a better way to merge CWRC and Swift hashes into an output format
+  # for each cwrc object
+  cwrc_objs&.each do |cwrc_obj|
+    cwrc_pid = cwrc_obj['pid'].to_s
+    cwrc_mtime = cwrc_obj['timestamp'].to_s
+    swift_id = cwrc_pid.dup
+    # account for the CWRC PID ':' replaced with "_" in the Swift ID
+    swift_id.sub! ':', '_'
+
+    if swift_objs.key?(swift_id)
+      swift_id = swift_id
```

Comment:
This doesn't create a new variable in this scope, this is just a no-op assignment to itself.